src/core/lib/surface/server.cc

```diff
@@ -1433,30 +1448,40 @@ static grpc_call_error queue_call_request(grpc_server* server, size_t cq_idx,
       rm = &rc->data.registered.method->matcher;
       break;
   }
-  if (gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link)) {
-    /* this was the first queued request: we need to lock and start
-       matching calls */
-    gpr_mu_lock(&server->mu_call);
-    while ((calld = rm->pending_head) != nullptr) {
-      rc = reinterpret_cast<requested_call*>(
-          gpr_locked_mpscq_pop(&rm->requests_per_cq[cq_idx]));
-      if (rc == nullptr) break;
-      rm->pending_head = calld->pending_next;
-      gpr_mu_unlock(&server->mu_call);
-      if (!gpr_atm_full_cas(&calld->state, PENDING, ACTIVATED)) {
-        // Zombied Call
-        GRPC_CLOSURE_INIT(
-            &calld->kill_zombie_closure, kill_zombie,
-            grpc_call_stack_element(grpc_call_get_call_stack(calld->call), 0),
-            grpc_schedule_on_exec_ctx);
-        GRPC_CLOSURE_SCHED(&calld->kill_zombie_closure, GRPC_ERROR_NONE);
-      } else {
-        publish_call(server, calld, cq_idx, rc);
-      }
-      gpr_mu_lock(&server->mu_call);
-    }
+
+  // Fast path: if there is no pending request to be processed, immediately
+  // return.
+  if (!gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link) ||
+      // Note: pending_head is written using order_release in publish_new_rpc().
+      //       Here we are reading the pointer without holding the sever call
+      //       mutex. The reset of stores and loads in this function are under
+      //       server call mutex and can use order_relaxed.
+      rm->pending_head.load(std::memory_order_acquire) == nullptr) {
```

Comment:
I believe there is a race here with line 582 where we might miss that pending_head is actually in the process of being set. Is that an okay race to have or is there something that fundamentally prevents this race from occurring?

Reply:
Thank you @hcaseyal. Yes, that's right, but that already exists even with the  existing version that's using the lock.

For example, suppose we have two threads: T1 and T2.
T1 is at publish_new_rpc() line 559 right before locking the mutex. T2 is in queue_call_request() line 1451 and adds the first call. If T2 wins locking the mutex, it would read a nullptr from pending_head and immediately return. 

With that, I think we are preserving existing behavior. Would that address your question?